### DISCUSSION TITLE
Question about Virtualization framework

### DISCUSSION THREAD

**Discussion by alexellis** [utmapp/UTM#5604] [Development]
Hi, I was looking to start playing with the Virtualization framework and Swift so thought you may be able to answer a question or two. No worries if not.

Is there a better resource than the very sparse Apple docs?

Have you found a way to send key strokes or commands to a MacOS VM through Virtualization framework from your own program?

I.e. let's say you've restored from a disk image, and see "Hello Bonjour" on the screen, have you been able to send a keystroke to walk through the installer?

Post install, I assume you can just enable SSH / VNC or install your own agent for communication.

Thanks 👍

**Comment by osy** [comment]
1. No, Apple’s same code is probably the best resource.
2. In theory you can send NSEvents to the VZVirtualMachineView. There’s probably also some private APIs you can use if you reverse it.
3. I’m pretty sure in private APIs you can enable VNC.
